File: Android/shared/src/main/kotlin/shared/dto/objects/Salon.kt
@file:UseSerializers(UUIDSerializer::class, URISerializer::class)

package shared.dto.objects

import kotlinx.serialization.Serializable
import kotlinx.serialization.UseSerializers
import shared.dto.enums.SalonType
import shared.dto.primitives.Address
import shared.dto.protocols.Parametable
import shared.dto.protocols.Responsable
import shared.serializers.URISerializer
import shared.serializers.UUIDSerializer
import java.net.URI
import java.util.*

object Salon {
    data object Parameters {
        @Serializable
        data class Create(
            val name: String,
            val type: SalonType,
            val logo: URI?,
            val timeZoneId: String,
            val localeId: String,
            val description: String?,
            val timetable: Timetable.Parameters.Create.Pattern,
            val address: Address,
            var contacts: List<Contact.Parameters.Create>,
        ) : Parametable()

        @Serializable
        data class Patch(
            val name: String?,
            val type: SalonType?,
            val description: String?,
            val logo: URI?,
        ) : Parametable()
    }

    data object Responses {
        @Serializable
        data class Full(
            var id: UUID,
            var name: String,
            var type: SalonType,
            var description: String?,
            var logo: URI? = null,
            var address: Address,
            var isActive: Boolean,
            var canEdit: Boolean = false,
            var isFavorite: Boolean = false,
            var localeId: String,
            var timeZoneId: String
        ) : Responsable

        @Serializable
        data class Partial(
            var id: UUID,
            var name: String,
            var type: SalonType,
            var logo: URI? = null,
            var address: Address,
            var isFavorite: Boolean = false,
        ) : Responsable
    }
}

File: Node/src/dto/objects/Salon.ts
import { Address } from "../primitives/Address"
import { SalonType } from "../enums/SalonType"
import { Timetable } from "./Timetable"
import { Contact } from "./Contact"
import { UUID } from "../tsPrimitives/UUID"
import { IsLocale, IsTimeZone, ValidateNested } from "class-validator"

export namespace Salon {

    export namespace Parameters {
    
        export class Create {
            name: string
            type: SalonType
            logo?: URL
            description?: string

            @IsTimeZone()
            timeZoneId: string

            @IsLocale()
            localeId: string

            @ValidateNested()
            timetable: Timetable.Parameters.Create.Pattern

            @ValidateNested()
            address: Address

            @ValidateNested()
            contacts: Contact.Parameters.Create[]

            constructor(
                name: string,
                type: SalonType,
                timeZoneId: string,
                localeId: string,
                timetable: Timetable.Parameters.Create.Pattern,
                address: Address,
                contacts: Contact.Parameters.Create[],
                logo?: URL,
                description?: string,
            ) {
                this.name = name
                this.type = type
                this.logo = logo
                this.timeZoneId = timeZoneId
                this.localeId = localeId
                this.description = description
                this.timetable = timetable
                this.address = address
                this.contacts = contacts
            }
        }

        export class Patch {
            name?: string
            type?: SalonType
            description?: string
            logo?: URL

            constructor(
                name?: string,
                type?: SalonType,
                description?: string,
                logo?: URL,
            ) {
                this.name = name
                this.type = type
                this.description = description
                this.logo = logo
            }
        }
    }

    export namespace Responses {

        export class Full {
            @ValidateNested()
            id: UUID
            
            name: string
            type: SalonType
            description?: string
            logo?: URL
            isActive: boolean
            canEdit: boolean
            isFavorite: boolean
            
            @ValidateNested()
            address: Address

            @IsLocale()
            localeId: string

            @IsTimeZone()
            timeZoneId: string

            constructor(
                id: UUID,
                name: string,
                type: SalonType,
                address: Address,
                isActive: boolean,
                canEdit: boolean,
                isFavorite: boolean,
                localeId: string,
                timeZoneId: string,
                description?: string,
                logo?: URL,
            ) {
                this.id = id
                this.name = name
                this.type = type
                this.description = description
                this.logo = logo
                this.address = address
                this.isActive = isActive
                this.canEdit = canEdit
                this.isFavorite = isFavorite
                this.localeId = localeId
                this.timeZoneId = timeZoneId
            }
        }

        export class Partial {
            @ValidateNested()
            id: UUID
            
            @ValidateNested()
            address: Address
            
            name: string
            type: SalonType
            logo?: URL
            isFavorite: boolean

            constructor(
                id: UUID,
                name: string,
                type: SalonType,
                address: Address,
                isFavorite: boolean,
                logo?: URL, 
            ) {
                this.id = id
                this.name = name
                this.type = type
                this.address = address
                this.isFavorite = isFavorite
                this.logo = logo
            }
        }
    }
} 